contracts/ChainlinkPriceFeedV3.sol

```diff
@@ -0,0 +1,197 @@
+import { IChainlinkPriceFeed } from "./interface/IChainlinkPriceFeed.sol";
+import { IChainlinkPriceFeedV3 } from "./interface/IChainlinkPriceFeedV3.sol";
+import { IPriceFeedUpdate } from "./interface/IPriceFeedUpdate.sol";
+import { BlockContext } from "./base/BlockContext.sol";
+import { CachedTwap } from "./twap/CachedTwap.sol";
+
+contract ChainlinkPriceFeedV3 is IChainlinkPriceFeedV3, IPriceFeedUpdate, BlockContext, CachedTwap {
+    using SafeMath for uint256;
+    using Address for address;
+
+    //
+    // STATE
+    //
+
+    uint8 internal immutable _decimals;
+    uint256 internal immutable _timeout;
+    uint256 internal _lastValidPrice;
+    uint256 internal _lastValidTimestamp;
+    AggregatorV3Interface internal immutable _aggregator;
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    constructor(
+        AggregatorV3Interface aggregator,
+        uint256 timeout,
+        uint80 twapInterval
+    ) CachedTwap(twapInterval) {
+        // CPF_ANC: Aggregator is not contract
+        require(address(aggregator).isContract(), "CPF_ANC");
+        _aggregator = aggregator;
+
+        _timeout = timeout;
+        _decimals = aggregator.decimals();
+    }
+
+    /// @notice anyone can help with updating
+    /// @dev keep this function for PriceFeedUpdater for updating, since multiple updates
+    ///      with the same timestamp will get reverted in CumulativeTwap._update()
+    function update() external override {
+        _cachePrice();
+
+        (bool isUpdated, ) = _cacheTwap(0, _lastValidPrice, _lastValidTimestamp);
+        // CPF_NU: not updated
+        require(isUpdated, "CPF_NU");
+    }
+
+    /// @inheritdoc IChainlinkPriceFeedV3
+    function cacheTwap(uint256 interval) external override {
+        _cachePrice();
+
+        _cacheTwap(interval, _lastValidPrice, _lastValidTimestamp);
+    }
+
+    //
+    // EXTERNAL VIEW
+    //
+
+    function getLastValidPrice() external view override returns (uint256) {
+        return _lastValidPrice;
+    }
+
+    function getLastValidTimestamp() external view override returns (uint256) {
+        return _lastValidTimestamp;
+    }
+
+    /// @inheritdoc IChainlinkPriceFeedV3
+    function getCachedTwap(uint256 interval) external view override returns (uint256) {
+        (uint256 latestValidPrice, uint256 latestValidTime) = _getCachePrice();
+
+        if (interval == 0) {
+            return latestValidPrice;
+        }
+
+        return _getCachedTwap(interval, latestValidPrice, latestValidTime);
+    }
+
+    function isTimedOut() external view override returns (bool) {
+        // Fetch the latest timstamp instead of _lastValidTimestamp is to prevent stale data
+        // when the update() doesn't get triggered.
+        (, uint256 lastestValidTimestamp) = _getCachePrice();
+        return lastestValidTimestamp > 0 && lastestValidTimestamp.add(_timeout) < _blockTimestamp();
+    }
```

Comment:
1. and 2. are both happen at the same time, right? 

This will only when chainlink is freezed at the very begining and `_lastValidTimestamp` and price will be 0. 

If this happens at a very begining, we might notice it and nothing will be plugged in to core contract.